src/plugin.js

```diff
@@ -44,129 +44,96 @@ const AD_STATES = [
+   * @return {Boolean}
+   */
+  const isInAdPlayback = () =>
+    !!this.ads && typeof this.ads === 'object' && AD_STATES.indexOf(this.ads.state) > -1;
+
+  /**
+   * Creates an event binder function of a given type.
+   *
+   * @param  {Boolean} isOne
+   *         Rather than delegating to the player's `one()` method, we want to
+   *         retain full control over when the listener is unbound (particularly
+   *         due to the ability for per-source behaviors to be toggled on and
+   *         off at will).
+   *
+   * @return {Function}
+   */
+  const createPerSrcBinder = (isOne) => {
+    return function(first, second) {
+
+      // Do not bind new listeners when per-source behaviors are disabled.
       if (this.perSourceBehaviors.disabled()) {
         return;
       }
 
-      if (changed || isOne) {
-        this.off(...args);
-      }
-
-      if (!changed) {
-        originalListener(...subargs);
+      const isTargetPlayer = arguments.length === 2;
+      const originalSrc = this.currentSrc();
+
+      // This array is the set of arguments to use for `on()` and `off()` methods
+      // of the player.
+      const args = [first];
+
+      // Make sure we bind here so that a GUID is set on the original listener
+      // and that it is bound to the proper context.
+      const originalListener = videojs.bind(
+        isTargetPlayer ? this : first,
+        arguments[arguments.length - 1]
+      );
+
+      // The wrapped listener `subargs` are the arguments passed to the original
+      // listener (i.e. the Event object and an additional data hash).
+      const wrappedListener = (...subargs) => {
+        const changed = this.currentSrc() !== originalSrc;
+
+        // Do not evaluate listeners if per-source behaviors are disabled.
+        if (this.perSourceBehaviors.disabled()) {
+          return;
+        }
+
+        if (changed || isOne) {
+          this.off(...args);
+        }
+
+        if (!changed) {
+          originalListener(...subargs);
+        }
+      };
+
+      // Make sure the wrapped listener and the original listener share a GUID,
+      // so that video.js properly removes event bindings when `off()` is passed
+      // the original listener!
+      wrappedListener.guid = originalListener.guid;
+
+      // If we are targeting a different object from the player, we need to include
+      // the second argument.
+      if (!isTargetPlayer) {
+        args.push(second);
       }
-    };
-
-    // Make sure the wrapped listener and the original listener share a GUID,
-    // so that video.js properly removes event bindings when `off()` is passed
-    // the original listener!
-    wrappedListener.guid = originalListener.guid;
-
-    // If we are targeting a different object from the player, we need to include
-    // the second argument.
-    if (!isTargetPlayer) {
-      args.push(second);
-    }
 
-    args.push(wrappedListener);
+      args.push(wrappedListener);
+      perSrcListeners.push(args);
```

Comment:
All these function moves were a result of wanting to keep this array private, so we could `push` to it without exposing it.